### PR TITLE
Feature Implementation: Adding Generic Plotter by scandate

### PR DESCRIPTION
**Pull request by bdorney**
## Description

Starting to address: https://github.com/cms-gem-daq-project/gem-plotting-tools/issues/33

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context

Right now if a particular plot of interest is desired a user has to make it by hand.  This feature seeks to address that.  The user provides a `tab` delimited text file `listOfScanDates.txt` of the form:

```
ChamberName scandate    <IndepVarName>
cName date    IndepVarVal1
cName date    IndepVarVal2
cName date    IndepVarVal3
...
...
cName date    IndepVarValN-1
cName date    IndepVarValN
```

And calls:

```
gemPlotter.py --anaType=trim --infilename=listOfScanDates.txt --vfatList=<comma separated list of integers> --strip=49 --branchName=<Name>
```

Then a plot of `branchName` vs `IndepVarName` from the scan dates specified, for each of the VFATs specified will be created.  Please note that `cName` need not be the same for each row (however then one would imagine that `IndepVar` is some expression derived from Layer & Slot...)!

## How Has This Been Tested?

By producing plots below.

### Screenshots (if appropriate):
Calling:
```
gemPlotter.py -ilistOfScanDates_Trim.txt -s10 -a --anaType=trimAna --branchName=trimDAC --axisMin=0 --axisMax=31
```
gemplotter_trimdac_vs_runno_vfat10_strip10_summary

Calling:
```
gemPlotter.py -ilistOfScanDates_Trim.txt -s10 -v10 --anaType=trimAna --branchName=trimDAC --axisMin=0 --axisMax=31
```
gemplotter_trimdac_vs_runno_vfat10_strip10

Calling:
```
gemPlotter.py -ilistOfScanDates_Trim.txt -a --anaType=trimAna --branchName=trimDAC --make2D --axisMin=-1 --axisMax=31
```
gemplotter_strip_vs_runno_obstrimdac_summary

Calling:
```
gemPlotter.py -ilistOfScanDates_Trim.txt -v10 --anaType=trimAna --branchName=trimDAC --make2D --axisMin=-1 --axisMax=31
```
gemplotter_strip_vs_runno_obstrimdac_vfat10

## Checklist:

- [X] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: mapping/chamberInfo.py [hunk 1]
```diff
@@ -1,55 +1,55 @@
...
+    #0:0x0,
+    #1:0x0,
+    #2:0x0,
+    #3:0x0,
+    #4:0x0,
+    #5:0x0,
+    #6:0x0,
+    #7:0x0,
+    #8:0x0,
+    #9:0x0
```
Comment (bdorney) [inline mapping/chamberInfo.py]: Shit this was a mistake.

Will fix it when I get back.

Do not merge.

File: .gitignore [hunk 2]
```diff
@@ -18,5 +18,6 @@
 #Don't ignore requirements.txt
 !requirements.txt
 
-#Don't ignore setup
+#Don't ignore select scripts
```
Comment (jsturdy) [inline .gitignore]: Why not `!*.sh`?

File: macros/gemPlotterAllChannels.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+for chan in {0..127}
+do
+    if [ $chan -eq 0 ]; then
+        echo gemPlotter.py -i$INFILE -s$chan -a --anaType=$ANATYPE --branchName=$BRANCHNAME --rootOpt=RECREATE
+        gemPlotter.py -i$INFILE -s$chan -a --anaType=$ANATYPE --branchName=$BRANCHNAME --rootOpt=RECREATE
+    else
+        echo gemPlotter.py -i$INFILE -s$chan -a --anaType=$ANATYPE --branchName=$BRANCHNAME --rootOpt=UPDATE
+        gemPlotter.py -i$INFILE -s$chan -a --anaType=$ANATYPE --branchName=$BRANCHNAME --rootOpt=UPDATE
+    fi
```
Comment (jsturdy) [inline macros/gemPlotterAllChannels.sh]: How does this behave if the user doesn't follow the syntax?
Are there "safe" defaults?
If not, you should add a `usage` function that checks the arguments, I should have a few examples laying around

File: .gitignore (hunk 2 above)
Comment (bdorney) [reply .gitignore]: I'm not sure we want just any script to be able to be added.  We have `!*.py` in the `.gitignore` but it seems like it takes some more thought to develop a python script than a shell script.  By allowing any shell script maybe we allow something to be committed by mistake.  Given that we have just two shell scripts at the moment I don't see the need to have `!*.sh`.

However, I'm not opposed to adding it.

File: macros/gemPlotterAllChannels.sh (hunk 3 above)
Comment (bdorney) [reply macros/gemPlotterAllChannels.sh]: No if the user doesn't follow the syntax all hell breaks loose (read "undefined behavior could occur). 

Could you point me to one of your examples?  I can then use that to make a `usage` function.

File: .gitignore (hunk 2 above)
Comment (jsturdy) [reply .gitignore]: OK, though from my perspective, if people aren't using `git commit -a` we're fine :)

File: macros/gemPlotterAllChannels.sh (hunk 3 above)
Comment (jsturdy) [reply macros/gemPlotterAllChannels.sh]: Something simple like (after your shebang):
```bash
usage() {
    echo 'Usage : ./gemPlotterAllChannels.sh <InFile> <anaType> <branchName>'
    echo '         InFile: description'
    echo '         anaType: description'
    echo '         branchName: description'
    kill -INT $$;
}

## could also add some option parsing with getopts or similar, but probably not necessary
inf=${1}
ana=${2}
branch=${3}
## then here you check the variables that are necessary:
if [ -z ${inf+x} ] || [ -z ${ana+x} ] || [ -z ${branch+x} ]
# or, without the assignment above
if [ -z ${3+x} ] || [ -z ${2+x} ] || [ -z ${3+x} ]
then
    usage
fi
```

**Comment by bdorney** [comment]
> Does the helper script correctly handle potential misuse, i.e., failing in the right way? I suppose gemPlotter will check the validity of the things that are passed in? (Or at least fail in a non-destructive way if not)
> - anaType is valid

There's a check by `gemPlotter.py` to make sure that `--anaType` is valid at [L185](https://github.com/bdorney/gem-plotting-tools/blob/30702a74c270be6d07329f6337ad243d5f3bd3eb/macros/gemPlotter.py#L185) and it will tell the user what values are supported if an incorrect value is given.

> - input file exists

If the argument of `--infilename` does not exist `gemPlotter.py` will raise an IOError exception, I will add a check in `gemPlotterAllChannels.sh` that this file exists before looping.

If a particular scandate given inside of `--infilename` does not exist then both `arbitraryPlotter()` and `arbitraryPlotter2D()` will raise an exception and tell the user they need to check the scandate which is responsible for the exception. 

> - branchname is valid

`root2array()` raises an exception for this but `arbitraryPlotter()` and `arbitraryPlotter2D()` catch this inside the check that the filename associated with the `scandate` exists.  I'll optimize.

**Comment by bdorney** [comment]
Added better exception handling.

# Case: input file does not exist
Handled in gemPlotter.py:
```
gemPlotter.py -iBadFileName -v10 --anaType=trimAna --branchName=trimDAC --make2D --axisMin=-1 --axisMax=31
BadFileName does not seem to exist
[Errno 2] No such file or directory: 'BadFileName'
```

Handled in gemPlotterAllChannels.py:
```
gemPlotterAllChannels.sh BadInputFile trimAna trimRange
Input file: BadInputFile not found
Please cross-check, exiting
```
Note, exits after one exception and does not spam 128 bad messages

# Case: anaType is bad
Handled in gemPlotter.py:

```
gemPlotter.py -ilistOfScanDates_Trim.txt -v10 --anaType=fail --branchName=trimDAC --make2D --axisMin=-1 --axisMax=31
Invalid analysis specificed, please select only from the list:
['trim', 'latency', 'scurveAna', 'threshold', 'trimAna', 'scurve']
```

Handled in gemPlotterAllChannels.py:
```
gemPlotterAllChannels.sh listOfScanDates_Trim.txt fail trimRange                      
gemPlotter.py -ilistOfScanDates_Trim.txt -s0 -a --anaType=fail --branchName=trimRange --rootOpt=RECREATE
Invalid analysis specificed, please select only from the list:
['trim', 'latency', 'scurveAna', 'threshold', 'trimAna', 'scurve']
```
Note, exits after one exception and does not spam 128 bad messages

# Case: branchName is bad
Handled in gemPlotter.py:
```
gemPlotter.py -ilistOfScanDates_Trim.txt -v10 --anaType=trimAna --branchName=fail --make2D --axisMin=-1 --axisMax=31
Branch fail not in TTree scurveFitTree of file /data/bigdisk/GEM-Data-Taking/GE11_QC8//GE11-VI-L-CERN-0002/trim/z4.000000//2017.09.04.20.12/SCurveData_Trimmed/SCurveFitData.root
Existing Branches are:
<ROOT.TBranch object ("vfatN") at 0x2fe02c0>
<ROOT.TBranch object ("vfatCH") at 0x2fefde0>
<ROOT.TBranch object ("ROBstr") at 0x2ff0300>
<ROOT.TBranch object ("mask") at 0x2ff0840>
<ROOT.TBranch object ("maskReason") at 0x2ff0d80>
<ROOT.TBranch object ("panPin") at 0x2ff12c0>
<ROOT.TBranch object ("trimRange") at 0x2ff1800>
<ROOT.TBranch object ("vthr") at 0x2ff1d40>
<ROOT.TBranch object ("trimDAC") at 0x2ff2280>
<ROOT.TBranch object ("threshold") at 0x2ff27c0>
<ROOT.TBranch object ("noise") at 0x2ff3e80>
<ROOT.TBranch object ("pedestal") at 0x2ff43a0>
<ROOT.TBranch object ("ped_eff") at 0x2ff48e0>
<ROOT.TBranchElement object ("scurve_h") at 0x2ff4e20>
<ROOT.TBranch object ("chi2") at 0x3029220>
<ROOT.TBranch object ("ndf") at 0x3029710>
<ROOT.TBranch object ("Nhigh") at 0x3029c30>
Please try again using one of the existing branches
```

Handled in gemPlotterAllChannels.py:
```
gemPlotterAllChannels.sh listOfScanDates_Trim.txt trimAna fail                                          
gemPlotter.py -ilistOfScanDates_Trim.txt -s0 -a --anaType=trimAna --branchName=fail --rootOpt=RECREATE
Branch fail not in TTree scurveFitTree of file /data/bigdisk/GEM-Data-Taking/GE11_QC8//GE11-VI-L-CERN-0002/trim/z4.000000//2017.09.04.20.12/SCurveData_Trimmed/SCurveFitData.root
Existing Branches are:
<ROOT.TBranch object ("vfatN") at 0x3e644d0>
<ROOT.TBranch object ("vfatCH") at 0x3e73ff0>
<ROOT.TBranch object ("ROBstr") at 0x3e74510>
<ROOT.TBranch object ("mask") at 0x3e74a50>
<ROOT.TBranch object ("maskReason") at 0x3e74f90>
<ROOT.TBranch object ("panPin") at 0x3e754d0>
<ROOT.TBranch object ("trimRange") at 0x3e75a10>
<ROOT.TBranch object ("vthr") at 0x3e75f50>
<ROOT.TBranch object ("trimDAC") at 0x3e76490>
<ROOT.TBranch object ("threshold") at 0x3e769d0>
<ROOT.TBranch object ("noise") at 0x3e78090>
<ROOT.TBranch object ("pedestal") at 0x3e785b0>
<ROOT.TBranch object ("ped_eff") at 0x3e78af0>
<ROOT.TBranchElement object ("scurve_h") at 0x3e79030>
<ROOT.TBranch object ("chi2") at 0x3ead430>
<ROOT.TBranch object ("ndf") at 0x3ead920>
<ROOT.TBranch object ("Nhigh") at 0x3eade40>
Please try again using one of the existing branches
```
Note, exits after one exception and does not spam 128 bad messages

# Case: scandate is bad
Handled in gemPlotter.py:
```
gemPlotter.py -ilistOfScanDates_Trim.txt -s0 -a --anaType=trimAna --branchName=trimRange --rootOpt=RECREATE
Filepath /data/bigdisk/GEM-Data-Taking/GE11_QC8//GE11-VI-L-CERN-0002/trim/z4.000000//9999.99.99.99.99 does not exist!
Please cross-check, exiting!
```

Handled in gemPlotterAllChannels.py:
```
gemPlotterAllChannels.sh listOfScanDates_Trim.txt trimAna trimRange
gemPlotter.py -ilistOfScanDates_Trim.txt -s0 -a --anaType=trimAna --branchName=trimRange --rootOpt=RECREATE
Filepath /data/bigdisk/GEM-Data-Taking/GE11_QC8//GE11-VI-L-CERN-0002/trim/z4.000000//9999.99.99.99.99 does not exist!
Please cross-check, exiting!
```
Note, exits after one exception and does not spam 128 bad messages

# Case: ChamberName is bad
Handled in gemPlotter.py:
```
gemPlotter.py -ilistOfScanDates_Trim.txt -s0 -a --anaType=trimAna --branchName=trimRange --rootOpt=RECREATE
Filepath /data/bigdisk/GEM-Data-Taking/GE11_QC8//FakeDetector/trim/z4.000000//2017.09.04.20.12 does not exist!
Please cross-check, exiting!
```

Handled in gemPlotterAllChannels.py:
```
gemPlotterAllChannels.sh listOfScanDates_Trim.txt trimAna trimRange
gemPlotter.py -ilistOfScanDates_Trim.txt -s0 -a --anaType=trimAna --branchName=trimRange --rootOpt=RECREATE
Filepath /data/bigdisk/GEM-Data-Taking/GE11_QC8//FakeDetector/trim/z4.000000//2017.09.04.20.12 does not exist!
Please cross-check, exiting!
```
Note, exits after one exception and does not spam 128 bad messages

# Rechecked Working Cases
Regenerated above screen shots of working plots using commands shown above.

Looks good.

**Comment by bdorney** [comment]
Error codes are now positive and follow pythonic example given above.

Performed same set of tests as shown, both to generate exceptions, and also plots.  Worked out in the end.

**Pull request merged**